nuxeo-ai-core/src/main/java/org/nuxeo/ai/functions/SaveEnrichmentFunction.java

```diff
@@ -41,30 +42,52 @@ public class SaveEnrichmentFunction extends AbstractEnrichmentConsumer {
     public void accept(EnrichmentMetadata metadata) {
         TransactionHelper.runInTransaction(() -> CoreInstance.doPrivileged(metadata.context.repositoryName, session -> {
             DocMetadataService docMetadataService = Framework.getService(DocMetadataService.class);
+
+            log.debug("Saving enrichment for document {}.", metadata.context.documentRef);
             DocumentModel doc = docMetadataService.saveEnrichment(session, metadata);
-            if (doc != null) {
-                try {
-                    if (!doc.isImmutable()) {
-                        log.debug("Saving enrichment for document {}.", doc.getId());
-                        session.saveDocument(doc);
-                    } else {
-                        log.error("Attempt to write into an Immutable Document Model id: {}, AI Model name {}",
-                                doc.getId(), metadata.getModelName());
-                    }
-                } catch (DocumentValidationException e) {
-                    log.warn("Failed to save document enrichment data for {}; error {}", metadata.context.documentRef,
-                            e.getMessage());
-                    if (log.isDebugEnabled()) {
-                        // log field violations
-                        List<ConstraintViolation> violations = e.getReport().asList();
-                        for (ConstraintViolation violation : violations) {
-                            log.debug("Constraint: {}, Value: {}", violation.getConstraint().getDescription(),
-                                    violation.getInvalidValue());
-                        }
+
+            if (doc == null) {
+                log.warn("Failed to save enrichment for document {}.", metadata.context.documentRef);
+                return null;
+            }
+
+            log.debug("Checking if the document is checked out and if a base version exists for the document {}.",
+                    doc.getId());
+
+            DocumentRef baseVersionRef = session.getBaseVersion(doc.getRef());
+
+            if (baseVersionRef == null && !doc.isCheckedOut()) {
+                log.warn(
+                        "Failed to save enrichment. The document is not checked out and no base version was found for document {}.",
+                        doc.getId());
+                return null;
+            }
+
+            if (doc.isImmutable()) {
+                log.error("Attempt to write into an Immutable Document Model id: {}, AI Model name {}", doc.getId(),
+                        metadata.getModelName());
```

Comment:
It will be interesting to understand why we are trying to update an immutable version or proxy ? 
This should be filtered upstream instead of computing enrichment that cannot be persisted.

Reply:
Thanks, @bdelbosc! Filtering upstream sounds like a good idea for a future enhancement. I think this could be a good additional check here though, so we could keep this code even with filtering.